services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCMapper.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oid4vc.issuance.mappers;
+
+import org.keycloak.Config;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.ProtocolMapper;
+import org.keycloak.protocol.oid4vc.OID4VCLoginProtocolFactory;
+import org.keycloak.protocol.oid4vc.model.VerifiableCredential;
+import org.keycloak.provider.ProviderConfigProperty;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Base class for OID4VC Mappers, to provide common configuration and functionality for all of them
+ *
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+public abstract class OID4VCMapper implements ProtocolMapper {
```

Comment:
Should this one also implement `OID4VCEnvironmentProviderFactory` to make sure that protocol mappers are available just for `oid4vc` protocol? I assume that it is not strictly needed, but at the same time, it probably won't hurt?

Reply:
Yes, I added it in [a65798d](https://github.com/keycloak/keycloak/pull/27931/commits/a65798dfaa126bc4edbcc254a77affa5c24816bc)